From 1f0f17a209e4f56add9107e1008be7f93cd1f2ee Mon Sep 17 00:00:00 2001
From: Ayoub Zaki <ayoub.zaki@embetrix.com>
Date: Mon, 27 Apr 2026 18:55:08 +0200
Subject: [PATCH 2/2] dm verity: add root_hash_sig_hex optional parameter

Add support for passing the root hash PKCS#7 signature as a hex-encoded
DER blob directly on the dm-verity table line via a new optional
parameter "root_hash_sig_hex".  This allows dm-verity root hash
signature verification when using dm-mod.create= on the kernel command
line, without requiring an initramfs to load the signature from a file
or keyring.

The hex string is decoded to binary at table load time and passed to
verify_pkcs7_signature() the same way the existing
root_hash_sig_key_desc path does.

Reference: https://www.lynx.com/blog/dm-verity-without-an-initramfs/

Upstream-Status: Inappropriate

Signed-off-by: Ayoub Zaki <ayoub.zaki@embetrix.com>
---
 drivers/md/dm-verity-target.c     | 11 ++++++
 drivers/md/dm-verity-verify-sig.c | 60 +++++++++++++++++++++++++++++++
 drivers/md/dm-verity-verify-sig.h | 17 ++++++++-
 3 files changed, 87 insertions(+), 1 deletion(-)

diff --git a/drivers/md/dm-verity-target.c b/drivers/md/dm-verity-target.c
index 60e74c0e02f1..d5d75e786bda 100644
--- a/drivers/md/dm-verity-target.c
+++ b/drivers/md/dm-verity-target.c
@@ -1194,6 +1194,17 @@ static int verity_parse_opt_args(struct dm_arg_set *as, struct dm_verity *v,
 				return r;
 			continue;
 
+		} else if (verity_verify_is_hex_sig_opt_arg(arg_name)) {
+			if (only_modifier_opts)
+				continue;
+			r = verity_verify_hex_sig_parse_opt_args(as, v,
+								 verify_args,
+								 &argc,
+								 arg_name);
+			if (r)
+				return r;
+			continue;
+
 		} else if (only_modifier_opts) {
 			/*
 			 * Ignore unrecognized opt, could easily be an extra
diff --git a/drivers/md/dm-verity-verify-sig.c b/drivers/md/dm-verity-verify-sig.c
index 2d397cbb8c5c..d6d26d95701e 100644
--- a/drivers/md/dm-verity-verify-sig.c
+++ b/drivers/md/dm-verity-verify-sig.c
@@ -30,6 +30,12 @@ bool verity_verify_is_sig_opt_arg(const char *arg_name)
 			    DM_VERITY_ROOT_HASH_VERIFICATION_OPT_SIG_KEY));
 }
 
+bool verity_verify_is_hex_sig_opt_arg(const char *arg_name)
+{
+	return (!strcasecmp(arg_name,
+			    DM_VERITY_ROOT_HASH_VERIFICATION_OPT_SIG_HEX));
+}
+
 static int verity_verify_get_sig_from_key(const char *key_desc,
 					struct dm_verity_sig_opts *sig_opts)
 {
@@ -104,6 +110,60 @@ int verity_verify_sig_parse_opt_args(struct dm_arg_set *as,
 	return 0;
 }
 
+int verity_verify_hex_sig_parse_opt_args(struct dm_arg_set *as,
+					 struct dm_verity *v,
+					 struct dm_verity_sig_opts *sig_opts,
+					 unsigned int *argc,
+					 const char *arg_name)
+{
+	struct dm_target *ti = v->ti;
+	const char *hex_sig;
+	size_t hex_len, sig_len;
+
+	if (sig_opts->sig) {
+		ti->error = DM_VERITY_VERIFY_ERR("signature already specified");
+		return -EINVAL;
+	}
+
+	if (!*argc) {
+		ti->error = DM_VERITY_VERIFY_ERR("hex signature value not specified");
+		return -EINVAL;
+	}
+
+	hex_sig = dm_shift_arg(as);
+	(*argc)--;
+
+	hex_len = strlen(hex_sig);
+	if (!hex_len || (hex_len & 1)) {
+		ti->error = DM_VERITY_VERIFY_ERR("invalid hex signature length");
+		return -EINVAL;
+	}
+
+	sig_len = hex_len / 2;
+	sig_opts->sig = kmalloc(sig_len, GFP_KERNEL);
+	if (!sig_opts->sig) {
+		ti->error = DM_VERITY_VERIFY_ERR("cannot allocate signature");
+		return -ENOMEM;
+	}
+
+	if (hex2bin(sig_opts->sig, hex_sig, sig_len)) {
+		ti->error = DM_VERITY_VERIFY_ERR("invalid hex signature");
+		kfree(sig_opts->sig);
+		sig_opts->sig = NULL;
+		return -EINVAL;
+	}
+
+	sig_opts->sig_size = sig_len;
+
+	v->signature_key_desc = kstrdup("hex_sig", GFP_KERNEL);
+	if (!v->signature_key_desc) {
+		ti->error = DM_VERITY_VERIFY_ERR("cannot allocate signature desc");
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
 /*
  * verify_verify_roothash - Verify the root hash of the verity hash device
  *			     using builtin trusted keys.
diff --git a/drivers/md/dm-verity-verify-sig.h b/drivers/md/dm-verity-verify-sig.h
index f36ea92127bf..4add735cfea7 100644
--- a/drivers/md/dm-verity-verify-sig.h
+++ b/drivers/md/dm-verity-verify-sig.h
@@ -10,6 +10,7 @@
 
 #define DM_VERITY_ROOT_HASH_VERIFICATION "DM Verity Sig Verification"
 #define DM_VERITY_ROOT_HASH_VERIFICATION_OPT_SIG_KEY "root_hash_sig_key_desc"
+#define DM_VERITY_ROOT_HASH_VERIFICATION_OPT_SIG_HEX "root_hash_sig_hex"
 
 struct dm_verity_sig_opts {
 	unsigned int sig_size;
@@ -18,16 +19,23 @@ struct dm_verity_sig_opts {
 
 #ifdef CONFIG_DM_VERITY_VERIFY_ROOTHASH_SIG
 
-#define DM_VERITY_ROOT_HASH_VERIFICATION_OPTS 2
+#define DM_VERITY_ROOT_HASH_VERIFICATION_OPTS 4
 
 int verity_verify_root_hash(const void *data, size_t data_len,
 			    const void *sig_data, size_t sig_len);
 bool verity_verify_is_sig_opt_arg(const char *arg_name);
+bool verity_verify_is_hex_sig_opt_arg(const char *arg_name);
 
 int verity_verify_sig_parse_opt_args(struct dm_arg_set *as, struct dm_verity *v,
 				    struct dm_verity_sig_opts *sig_opts,
 				    unsigned int *argc, const char *arg_name);
 
+int verity_verify_hex_sig_parse_opt_args(struct dm_arg_set *as,
+					 struct dm_verity *v,
+					 struct dm_verity_sig_opts *sig_opts,
+					 unsigned int *argc,
+					 const char *arg_name);
+
 void verity_verify_sig_opts_cleanup(struct dm_verity_sig_opts *sig_opts);
 
 #else
@@ -52,6 +60,13 @@ static inline int verity_verify_sig_parse_opt_args(struct dm_arg_set *as,
 	return -EINVAL;
 }
 
+static inline int verity_verify_hex_sig_parse_opt_args(struct dm_arg_set *as,
+			struct dm_verity *v, struct dm_verity_sig_opts *sig_opts,
+			unsigned int *argc, const char *arg_name)
+{
+	return -EINVAL;
+}
+
 static inline void verity_verify_sig_opts_cleanup(struct dm_verity_sig_opts *sig_opts)
 {
 }
-- 
2.34.1

